Add new experimental presentation mode for full-window video viewing

danae404 opened the pull request:
#### f35e3cd3da0099400e5cd75be791681152e8ec90
<pre>
Add new experimental presentation mode for full-window video viewing
<a href="https://bugs.webkit.org/show_bug.cgi?id=266241">https://bugs.webkit.org/show_bug.cgi?id=266241</a>
<a href="https://rdar.apple.com/119503825">rdar://119503825</a>

Reviewed by Tim Nguyen and Jer Noble.

This patch adds a new WebKit presentation mode, In-Window Fullscreen, that puts a video element
In a fullscreen mode that is limited to the application&apos;s window. This mode is put behind a WebKit internal feature flag. This patch also adds a new pseudo-class for user agents only to style the video element when in this presentation mode: -internal-in-window-full-screen.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/modern-media-controls/media/media-controller.js:
(MediaController.prototype.get isFullscreen):
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesInWindowFullScreenPseudoClass):
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::JSC_DEFINE_JIT_OPERATION):
(WebCore::SelectorCompiler::addPseudoClassType):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::willEnterFullscreen):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::enterFullscreen):
(WebCore::HTMLMediaElement::exitFullscreen):
(WebCore::HTMLMediaElement::willBecomeFullscreenElement):
(WebCore::HTMLMediaElement::willStopBeingFullscreenElement):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::toFullscreenMode):
(WebCore::HTMLVideoElement::toPresentationMode):
(WebCore::HTMLVideoElement::webkitSetPresentationMode):
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebCore/html/HTMLVideoElement.idl:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::enterFullScreenForElement):
* Source/WebCore/platform/graphics/MediaPlayerEnums.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::willEnterFullScreen):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
(WebKit::WebFullScreenManager::exitFullScreenForElement):
(WebKit::WebFullScreenManager::willEnterFullScreen):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.messages.in:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.cpp:
(WebKit::InjectedBundlePageFullScreenClient::enterFullScreenForElement):
(WebKit::InjectedBundlePageFullScreenClient::exitFullScreenForElement):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::enterFullScreenForElement):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationManager::supportsVideoFullscreen const):

Canonical link: <a href="https://commits.webkit.org/273923@main">https://commits.webkit.org/273923@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ebe64ca5003bbfe2bd4bf4a07b9196771ab2a586

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36921 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15850 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39187 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39546 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33042 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38207 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18561 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12972 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31582 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37483 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13495 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32600 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11679 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11770 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32886 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40799 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/30930 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33589 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33231 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37597 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/36659 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11983 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9794 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35723 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/languages (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13636 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/43480 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8402 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12365 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8975 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12892 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/css/SelectorCheckerTestFunctions.h:
```diff
@@ -456,6 +456,17 @@ ALWAYS_INLINE bool matchesFullScreenControlsHiddenPseudoClass(const Element& ele
+    if (!is<HTMLMediaElement>(element))
+        return false;
+
+    return dynamicDowncast<HTMLMediaElement>(element)->fullscreenMode() == HTMLMediaElementEnums::VideoFullscreenModeInWindow;
```
```suggestion
if (auto* mediaElement = dynamicDowncast<HTMLMediaElement>(element); mediaElement && mediaElement->fullscreenMode() == HTMLMediaElementEnums::VideoFullscreenModeInWindow)
    return true;
    
  return false;
```

rr-codes commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -516,7 +516,10 @@ bool FullscreenManager::willEnterFullscreen(Element& element)
+    if (is<HTMLMediaElement>(element))
+        dynamicDowncast<HTMLMediaElement>(element)->willBecomeFullscreenElement(mode);
+    else
+        element.willBecomeFullscreenElement();
```
```suggestion
if (auto* mediaElement = dynamicDowncast<HTMLMediaElement>(element))
    mediaElement->willBecomeFullscreenElement(mode);
else
    element.willBecomeFullscreenElement();
```

rr-codes commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6974,14 +6974,14 @@ WEBCORE_EXPORT void HTMLMediaElement::setVideoFullscreenStandby(bool value)
 
-    if (m_videoFullscreenMode != VideoFullscreenModeStandard)
-        setFullscreenMode(VideoFullscreenModeStandard);
+    if (!(m_videoFullscreenMode == VideoFullscreenModeStandard || m_videoFullscreenMode == VideoFullscreenModeInWindow))
```
since you’re doing `== standard || == inWindow` a lot, maybe it could be refactored into a static method that’s like `isStandardOrInWindow` or something like that?

annevk replied:
This still seems like a good idea.

rr-codes commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -224,7 +224,13 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element)
+    if (mode == WebCore::HTMLMediaElementEnums::VideoFullscreenModeInWindow) {
+        willEnterFullScreen(mode);
+        didEnterFullScreen();
+        m_isInInWindowFullScreenMode = true;
```
“isInIn” kinda sounds a bit weird

danae404 replied:
I definitely agree haha. I'll try to think of something better

danae404 replied:
I changed it to m_inWindowFullScreenMode

nt1m commented on Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
```diff
@@ -85,6 +85,7 @@ fullscreen
 -webkit-full-screen-ancestor
 -webkit-full-screen-document
 -webkit-full-screen-controls-hidden
+-webkit-in-window-full-screen
```
Please prefix with `-internal-` instead of `-webkit-` and make this UA sheet only, see: https://searchfox.org/wubkat/rev/3943142d2ef1777fc8977e6158eb10e053c3f8ca/Source/WebCore/css/parser/CSSSelectorParser.cpp#783-784

danae404 replied:
Got it! Thanks

nt1m reviewed (changes requested):
Requesting changes until making the pseudo class internal has been addressed.

nt1m commented on Source/WebCore/css/CSSPseudoSelectors.json:
```diff
@@ -49,6 +49,10 @@
         "-internal-html-document": {},
+        "-internal-in-window-full-screen": {
+            "conditional": "ENABLE(FULLSCREEN_API)",
+            "status": "non-standard"
```
You can remove "status": "none-standard", it's internal so that's implicit

annevk commented on Source/WebCore/css/SelectorCheckerTestFunctions.h:
```diff
@@ -432,6 +432,17 @@ ALWAYS_INLINE bool matchesFullscreenDocumentPseudoClass(const Element& element)
+    if (auto* mediaElement = dynamicDowncast<HTMLMediaElement>(element); mediaElement && mediaElement->fullscreenMode() == HTMLMediaElementEnums::VideoFullscreenModeInWindow)
+        return true;
+
+    return false;
```
You want to use RefPtr and there's a more concise way to write this:
```suggestion
    RefPtr mediaElement = dynamicDowncast<HTMLMediaElement>(element);
    return mediaElement && mediaElement->fullscreenMode() == HTMLMediaElementEnums::VideoFullscreenModeInWindow;
```

danae404 replied:
Much more concise! Thank you

cdumez replied:
You don't need a RefPtr there though.

annevk commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -528,7 +529,10 @@ bool FullscreenManager::willEnterFullscreen(Element& element)
     ASSERT(page()->settings().fullScreenEnabled());
 
-    element.willBecomeFullscreenElement();
+    if (auto* mediaElement = dynamicDowncast<HTMLMediaElement>(element))
```
RefPtr

danae404 committed:
Add new experimental presentation mode for full-window video viewing
https://bugs.webkit.org/show_bug.cgi?id=266241
rdar://119503825

Reviewed by Tim Nguyen and Jer Noble.

This patch adds a new WebKit presentation mode, In-Window Fullscreen, that puts a video element
In a fullscreen mode that is limited to the application's window. This mode is put behind a WebKit internal feature flag. This patch also adds a new pseudo-class for user agents only to style the video element when in this presentation mode: -internal-in-window-full-screen.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/modern-media-controls/media/media-controller.js:
(MediaController.prototype.get isFullscreen):
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesInWindowFullScreenPseudoClass):
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::JSC_DEFINE_JIT_OPERATION):
(WebCore::SelectorCompiler::addPseudoClassType):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::willEnterFullscreen):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::enterFullscreen):
(WebCore::HTMLMediaElement::exitFullscreen):
(WebCore::HTMLMediaElement::willBecomeFullscreenElement):
(WebCore::HTMLMediaElement::willStopBeingFullscreenElement):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::toFullscreenMode):
(WebCore::HTMLVideoElement::toPresentationMode):
(WebCore::HTMLVideoElement::webkitSetPresentationMode):
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebCore/html/HTMLVideoElement.idl:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::enterFullScreenForElement):
* Source/WebCore/platform/graphics/MediaPlayerEnums.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::willEnterFullScreen):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
(WebKit::WebFullScreenManager::exitFullScreenForElement):
(WebKit::WebFullScreenManager::willEnterFullScreen):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.messages.in:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.cpp:
(WebKit::InjectedBundlePageFullScreenClient::enterFullScreenForElement):
(WebKit::InjectedBundlePageFullScreenClient::exitFullScreenForElement):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::enterFullScreenForElement):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationManager::supportsVideoFullscreen const):

Canonical link: https://commits.webkit.org/273923@main
